src/streamed/state.ts

```diff
@@ -165,135 +167,256 @@ export class AuctionState {
+    if (this.options.strict) {
+      for (const [user, { balances }] of this.accounts.entries()) {
+        for (const [token, balance] of balances.entries()) {
+          assert(balance >= BigInt(0), `user ${user} token ${token} balance is negative`)
+        }
+      }
+    }
+  }
+
+  /**
+   * Applies a token listing event and adds a token to the account state.
+   *
+   * @throws
+   * In strict mode, throws if the token has already been listed.
+   */
+  private applyTokenListing({ id, token }: Event<BatchExchange, "TokenListing">) {
+    if (this.options.strict) {
+      assert(
+        this.tokens.length === parseInt(id),
+        `token ${token} with ID ${id} added as the ${this.tokens.length} token`,
+      )
+    }
+
+    this.tokens.push(token)
+  }
+
+  /**
+   * Applies a trade event to the auction state.
    */
-  private addToken(id: TokenId, addr: Address): void {
-    this.tokens.set(id, addr)
+  private applyTrade(trade: Event<BatchExchange, "Trade">) {
+    this.updateBalance(
+      trade.owner,
+      parseInt(trade.sellToken),
+      amount => amount - BigInt(trade.executedSellAmount),
+    )
+    this.updateOrderRemainingAmount(
+      trade.owner,
+      parseInt(trade.orderId),
+      amount => amount - BigInt(trade.executedSellAmount),
+    )
+    this.updateBalance(
+      trade.owner,
+      parseInt(trade.buyToken),
+      amount => amount + BigInt(trade.executedBuyAmount),
+    )
+  }
+
+  /**
+   * Applies a trade reversion event to the auction state.
+   */
+  private applyTradeReversion(trade: Event<BatchExchange, "TradeReversion">) {
+    this.updateBalance(
+      trade.owner,
+      parseInt(trade.sellToken),
+      amount => amount + BigInt(trade.executedSellAmount),
+    )
+    this.updateOrderRemainingAmount(
+      trade.owner,
+      parseInt(trade.orderId),
+      amount => amount + BigInt(trade.executedSellAmount),
+    )
+    this.updateBalance(
+      trade.owner,
+      parseInt(trade.buyToken),
+      amount => amount - BigInt(trade.executedBuyAmount),
+    )
+  }
+
+  /**
+   * Applies a withdraw event to the auction state.
+   *
+   * @throws
+   * In strict mode, throws if the withdrawing user's balance would be overdrawn
+   * as a result of this event.
+   */
+  private applyWithdraw(
+    { user, token, amount: withdrawAmount }: Event<BatchExchange, "Withdraw">,
+  ) {
+    const tokenAddr = this.tokenAddr(token)
+    const newBalance = this.updateBalance(
+      user, token, amount => amount - BigInt(withdrawAmount),
+    )
+
+    if (this.options.strict) {
+      assert(
+        newBalance >= BigInt(0),
+        `overdrew user ${user} token ${token} balance`,
+      )
+    }
+
+    this.account(user).pendingWithdrawals.delete(tokenAddr)
+  }
+
+  /**
+   * Applies a withdraw request event to the auction state.
+   *
+   * @throws
+   * In strict mode, throws if the withdraw request is placed in the over top of
+   * and existing unapplied request.
```

Comment:
I think we have some typos here. Also, I am not sure why a withdraw request placed over top of an existing should throw. I believe throw contract would overwrite the existing one.

Reply:
:see_no_evil: 

In the smart contract, a withdraw is done to clear the previous withdraw request if it has already matured. If we ever get a withdraw request that would effectively overwrite a previous one, it means that we missed a withdraw event somewhere.